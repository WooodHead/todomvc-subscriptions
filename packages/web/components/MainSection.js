/* global window */
/* eslint jsx-a11y/label-has-for: 0, jsx-a11y/click-events-have-key-events: 0, jsx-a11y/no-noninteractive-element-interactions: 0 */
// @flow
import React, { Component } from 'react';
import { compose } from 'react-apollo';
import TodoItem from './TodoItem';
import Footer from './Footer';
import {
  withTodos,
  withToggleAll,
  type Todos,
  type ToggleAll,
  type SubscribeTodoChanged,
} from '../modules/todos/hoc';
import { withFilter } from '../modules/filter/hoc';
import {
  SHOW_ALL,
  SHOW_COMPLETED,
  SHOW_ACTIVE,
  type Filter,
} from '../modules/filter/constants';

type Props = {|
  todos: Todos,
  toggleAll: ToggleAll,
  subscribeTodoChanged: SubscribeTodoChanged,
  filter: {
    filter: Filter,
  },
|};

const TODO_FILTERS = {
  [SHOW_ALL]: () => true,
  [SHOW_ACTIVE]: todo => !todo.completed,
  [SHOW_COMPLETED]: todo => todo.completed,
};

export class MainSection extends Component<Props> {
  componentDidMount() {
    const clientID = window.localStorage.getItem('apollo-client-id');
    this.props.subscribeTodoChanged(clientID);
  }

  render() {
    const { todos: { todos = [] }, toggleAll, filter: { filter } } = this.props;
    const completedTodos = todos.filter(TODO_FILTERS[SHOW_COMPLETED]);
    const completedCount = completedTodos.length;
    const todoLength = todos.length;
    const isAllCompleted = completedCount === todoLength;

    return (
      <section className="main">
        <span>
          <input
            className="toggle-all"
            type="checkbox"
            checked={isAllCompleted}
          />
          <label onClick={() => toggleAll(!isAllCompleted)} />
        </span>
        <ul className="todo-list">
          {todos
            .filter(TODO_FILTERS[filter])
            .map(todo => <TodoItem key={todo.id} todo={todo} />)}
        </ul>
        <Footer
          filter={filter}
          todoLength={todoLength}
          completedTodos={completedTodos}
        />
      </section>
    );
  }
}

export default compose(withTodos, withFilter, withToggleAll)(MainSection);
