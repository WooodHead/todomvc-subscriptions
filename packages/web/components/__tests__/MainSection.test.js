/* global window */

import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { MainSection } from '../MainSection';

it('should shallow render <MainSection /> correctly', () => {
  window.localStorage = {
    getItem: jest.fn(),
  };
  const wrapper = shallow(
    <MainSection
      todos={{ todos: [] }}
      toggleAll={() => {}}
      subscribeTodoChanged={() => {}}
      filter={{ filter: 'SHOW_ALL' }}
    />,
  );

  expect(toJson(wrapper)).toMatchSnapshot();
  expect(window.localStorage.getItem).toMatchSnapshot();
});
