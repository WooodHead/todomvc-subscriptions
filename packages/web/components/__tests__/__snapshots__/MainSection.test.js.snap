// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should shallow render <MainSection /> correctly 1`] = `
<section
  className="main"
>
  <span>
    <input
      checked={true}
      className="toggle-all"
      type="checkbox"
    />
    <label
      onClick={[Function]}
    />
  </span>
  <ul
    className="todo-list"
  />
  <withUpdateFilter(withDeleteTodos(Footer))
    completedTodos={Array []}
    filter="SHOW_ALL"
    todoLength={0}
  />
</section>
`;

exports[`should shallow render <MainSection /> correctly 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "apollo-client-id",
    ],
  ],
}
`;
